From acfe832d763fd45185e02b750a9b175188ac3766 Mon Sep 17 00:00:00 2001
From: Brandon Maier <brandon.maier@gmail.com>
Date: Thu, 21 Nov 2024 18:33:57 -0600
Subject: [PATCH 1/3] tests: support admin check on Windows

The tests fail on Windows with the following

> ___________ ERROR collecting tests/source_file/test_source_file.py ____________
> tests\source_file\test_source_file.py:24: in <module>
>     class testOSError(unittest.TestCase):
> tests\source_file\test_source_file.py:39: in testOSError
>     @unittest.skipIf("SUDO_UID" in os.environ.keys() or os.geteuid() == 0, "We are root. Root always has permissions so test will fail.")
> E   AttributeError: module 'os' has no attribute 'geteuid'

Catch the AttributeError and attempt to check for admin when on Windows

Signed-off-by: Brandon Maier <brandon.maier@gmail.com>
---
 tests/source_file/test_source_file.py |  2 +-
 tests/utils.py                        | 13 +++++++++++++
 tests/vsg/test_vsg.py                 |  2 +-
 3 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/tests/source_file/test_source_file.py b/tests/source_file/test_source_file.py
index 49e680d8..7b53f4bd 100644
--- a/tests/source_file/test_source_file.py
+++ b/tests/source_file/test_source_file.py
@@ -36,7 +36,7 @@ def test_file_not_found(self):
 
         self.assertEqual(exit_status, 1)
 
-    @unittest.skipIf("SUDO_UID" in os.environ.keys() or os.geteuid() == 0, "We are root. Root always has permissions so test will fail.")
+    @unittest.skipIf(utils.is_user_admin(), "We are root. Root always has permissions so test will fail.")
     def test_file_no_permission(self):
         sNoPermissionTempFile = os.path.join(self._tmpdir.name, sNoPermissionFile)
 
diff --git a/tests/utils.py b/tests/utils.py
index e4ace646..ebdaa972 100644
--- a/tests/utils.py
+++ b/tests/utils.py
@@ -1,4 +1,5 @@
 # -*- coding: utf-8 -*-
+import ctypes
 import os
 import pprint
 import re
@@ -128,3 +129,15 @@ def replace_total_count_summary(lOutput):
 
 def replace_token(lOutput, src, dst):
     return [line.replace(src, dst) for line in lOutput]
+
+
+def is_user_admin():
+    if "SUDO_UID" in os.environ.keys():
+        return True
+
+    try:
+        return os.getuid() == 0
+    except AttributeError:
+        pass
+
+    return ctypes.windll.shell32.IsUserAnAdmin() == 1
diff --git a/tests/vsg/test_vsg.py b/tests/vsg/test_vsg.py
index db756bb1..f1b8ae09 100644
--- a/tests/vsg/test_vsg.py
+++ b/tests/vsg/test_vsg.py
@@ -343,7 +343,7 @@ def test_missing_configuration_file(self):
 
         self.assertEqual(iExitStatus, 1)
 
-    @unittest.skipIf("SUDO_UID" in os.environ.keys() or os.geteuid() == 0, "We are root. Root always has permissions so test will fail.")
+    @unittest.skipIf(utils.is_user_admin(), "We are root. Root always has permissions so test will fail.")
     def test_no_permission_configuration_file(self):
         sNoPermissionFile = os.path.join(self._tmpdir.name, "no_permission.yml")
         pathlib.Path(sNoPermissionFile).touch(mode=0o222, exist_ok=True)
-- 
2.47.0

